### PR TITLE
Bug 1881082: Expose etcd raft term as a metric

### PR DESCRIPTION
**Pull request by ironcladlou**
Piggyback on the member health checking facility to expose the current
raft term per member as a Prometheus metric to enable more granular
performance analysis.

### CODE REVIEW

**Comment by ironcladlou** [comment]
Here's what it looks like in a cluster I've got which has experienced massive churn:

```
# HELP etcd_raft_terms_total Number of etcd raft terms as observed by each member.
# TYPE etcd_raft_terms_total counter
etcd_raft_terms_total{member="dmace-71ad-nwdm7-master-0.c.openshift-gce-devel.internal"} 22
etcd_raft_terms_total{member="dmace-71ad-nwdm7-master-1.c.openshift-gce-devel.internal"} 22
etcd_raft_terms_total{member="dmace-71ad-nwdm7-master-2.c.openshift-gce-devel.internal"} 22
```

**Comment by ironcladlou** [comment]
Although I'm not seeing them being scraped by Prometheus in my cluster, so I might be missing something in terms of registration, etc.

/hold

**Comment by ironcladlou** [comment]
Already spent way too much time trying to figure out why modules are screwy in the CI build — switched to the deprecated `RawMustRegister()` function which bypasses the need to import semver since we have no use for the weird k8s metrics wrapper junk anyway.

**Comment by ironcladlou** [comment]
From a cluster-bot cluster running this PR:

Screen Shot 2020-09-15 at 10 59 24 AM

**Comment by hexfusion** [comment]
I think in general this is a net gain in the context of historical reference. Raft terms are persisted to state thus the counter is immutable so to speak vs the current metric. As a followup, I would like to see the term added to events where reasonable as raft terms are included in all API responses thus very available.. Another benefit of adding the term to events is we can track before prom starts.

/lgtm

File: pkg/etcdcli/health.go [hunk 1]
```diff
@@ -7,10 +7,29 @@ import (
...
 	"go.etcd.io/etcd/etcdserver/api/v3rpc/rpctypes"
 	"go.etcd.io/etcd/etcdserver/etcdserverpb"
+	"k8s.io/component-base/metrics/legacyregistry"
 )
 
+func init() {
+	legacyregistry.RawMustRegister(raftTerms)
+}
+
+const raftTermsMetricName = "etcd_raft_terms_total"
```
Comment (hexfusion) [inline pkg/etcdcli/health.go:20]: nit: should we rename this?

`etcd_debugging_raft_terms_total`

Per upstream

```
Note that any etcd_debugging_* metrics are experimental and subject to change.
```
This gives us the option to change remove if we feel so fit. I think any new metric should be considered experimental through one release.

Comment (ironcladlou) [reply pkg/etcdcli/health.go]: Good call, fixed

**Pull request merged**